### PR TITLE
Implemented QueuedDataManager

### PR DESCRIPTION
**Pull request by udeyrishi**
Implemented QueuedDataManager. There are some minor API changes, but I fixed all the places that were affected. 

QueuedDataManager guarantees:
* All the write and delete operations will eventually go through, no matter what the current internet state is. The requests are persisted, so that they are still sent even if the app restarts.
* All the operations will be sent in the exact same order as they were queued. This guarantees data integrity for operations that work on the same remote object.
* All the write and delete operations will be performed on a different thread. So these methods can be called on the UI thread safely, even though network is accessed.

To get an all configured QDM instance, you can go <pre><code>TradeApp.getInstance().createDataManager(useExplicitExposeAnnotation: boolean)</code></pre>.

### CODE REVIEW

**Comment by udeyrishi** [comment]
2 QDM tests are currently throwing NotImplementedException, as they need to restart the app in the test logic, and I'm not sure how to do it

File: T03/app/src/main/java/ca/ualberta/cmput301/t03/datamanager/CachedDataManager.java [hunk 1]
```diff
@@ -125,7 +108,7 @@ public boolean keyExists(DataKey key) throws IOException {
     public <T> void writeData(DataKey key, T obj, Type typeOfT) throws IOException {
         if (isOperational()) {
             innerManager.writeData(key, obj, typeOfT);
-            cachingDataManager.writeData(convertToCacheKey(key), obj, typeOfT);
+            writeToCache(key, obj, typeOfT);
```
Comment (MEGASTACK) [inline T03/app/src/main/java/ca/ualberta/cmput301/t03/datamanager/CachedDataManager.java]: lol gotta love java generics...

**Comment by MEGASTACK** [comment]
When performing certain operations that require network  offline, throwing servicenotavailableexception - which is a runtime exception, and therefore does not need to be explicitly caught - is probably not a good idea. It might be wise to change this exception to one that must be explicitly caught.

**Comment by udeyrishi** [comment]
I'll create a work item for it

**Comment by udeyrishi** [comment]
See issue #69

**Pull request merged**